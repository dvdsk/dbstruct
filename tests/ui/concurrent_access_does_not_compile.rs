use dbstruct::dbstruct;

#[dbstruct(db=sled)]
struct Test {
    list: Vec<u8>,
}

// SHOULD NOT COMPILE
fn main() {
    let test = Test::new("").unwrap();
    std::thread::scope(|s| {
        s.spawn(|| {
            test.list().push(&5).unwrap();
        });
        s.spawn(|| {
            test.list().push(&6).unwrap();
        });
    })
}
