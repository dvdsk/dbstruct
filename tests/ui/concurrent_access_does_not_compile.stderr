error[E0277]: `Cell<()>` cannot be shared between threads safely
  --> tests/ui/concurrent_access_does_not_compile.rs:12:17
   |
12 |           s.spawn(|| {
   |  ___________-----_^
   | |           |
   | |           required by a bound introduced by this call
13 | |             test.list().push(&5).unwrap();
14 | |         });
   | |_________^ `Cell<()>` cannot be shared between threads safely
   |
   = help: within `Test`, the trait `Sync` is not implemented for `Cell<()>`
   = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::RwLock`
note: required because it appears within the type `PhantomData<Cell<()>>`
  --> $RUST/core/src/marker.rs
   |
   | pub struct PhantomData<T: ?Sized>;
   |            ^^^^^^^^^^^
note: required because it appears within the type `Test`
  --> tests/ui/concurrent_access_does_not_compile.rs:4:8
   |
4  | struct Test {
   |        ^^^^
   = note: required for `&Test` to implement `Send`
note: required because it's used within this closure
  --> tests/ui/concurrent_access_does_not_compile.rs:12:17
   |
12 |         s.spawn(|| {
   |                 ^^
note: required by a bound in `Scope::<'scope, 'env>::spawn`
  --> $RUST/std/src/thread/scoped.rs
   |
   |     pub fn spawn<F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
   |            ----- required by a bound in this associated function
   |     where
   |         F: FnOnce() -> T + Send + 'scope,
   |                            ^^^^ required by this bound in `Scope::<'scope, 'env>::spawn`
